File: frontend/lib/business-info.ts
import type { BusinessInfoBlock } from "@/lib/types";

export const BUSINESS_INFO_TOPIC_ORDER: BusinessInfoBlock["topic"][] = [
  "hours",
  "location",
  "policy",
  "faq",
  "other",
];

export const DEFAULT_BUSINESS_INFO_BLOCKS: BusinessInfoBlock[] = BUSINESS_INFO_TOPIC_ORDER.map(
  (topic) => ({
    topic,
    content_he: "",
    content_en: "",
  }),
);

export function normalizeBusinessInfoBlocks(
  items: BusinessInfoBlock[],
): BusinessInfoBlock[] {
  const byTopic = new Map<BusinessInfoBlock["topic"], BusinessInfoBlock>();
  for (const item of items) {
    byTopic.set(item.topic, item);
  }
  return BUSINESS_INFO_TOPIC_ORDER.map(
    (topic) => byTopic.get(topic) ?? { topic, content_he: "", content_en: "" },
  );
}

File: backend/app/core/business_info.py
"""Business info helpers — one row per topic per tenant."""

from __future__ import annotations

from typing import TYPE_CHECKING, TypeVar

if TYPE_CHECKING:
    from app.core.schema import BusinessInfo

T = TypeVar("T")

BUSINESS_INFO_TOPIC_ORDER = ("hours", "location", "policy", "faq", "other")


def dedupe_business_info_payload(items: list[T], *, topic_getter) -> list[T]:
    """Keep the last item per topic in payload order."""
    by_topic: dict[str, T] = {}
    for item in items:
        by_topic[topic_getter(item)] = item
    return [by_topic[topic] for topic in BUSINESS_INFO_TOPIC_ORDER if topic in by_topic]


def dedupe_business_info_rows(rows: list["BusinessInfo"]) -> list["BusinessInfo"]:
    """Keep the newest row per topic (by updated_at, then created_at)."""
    by_topic: dict[str, BusinessInfo] = {}
    for row in rows:
        existing = by_topic.get(row.topic)
        if existing is None:
            by_topic[row.topic] = row
            continue
        row_stamp = (row.updated_at, row.created_at)
        existing_stamp = (existing.updated_at, existing.created_at)
        if row_stamp >= existing_stamp:
            by_topic[row.topic] = row
    return [by_topic[topic] for topic in BUSINESS_INFO_TOPIC_ORDER if topic in by_topic]
